src/gen/get_puzzle_and_solution.rs

```diff
@@ -5,24 +5,43 @@ use clvmr::op_utils::u64_from_bytes;
 use clvmr::sha2::{Digest, Sha256};
 use std::convert::AsRef;
 
-// TOOD: make this not recursive
+enum TreeOp {
+    SExp(NodePtr),
+    Cons,
+}
+
 // TODO: perhaps this should move to clvm_rs
-fn tree_hash(a: &Allocator, node: NodePtr) -> [u8; 32] {
-    match a.sexp(node) {
-        SExp::Atom(atom) => {
-            let mut sha256 = Sha256::new();
-            sha256.update(&[1_u8]);
-            sha256.update(a.buf(&atom));
-            sha256.finalize().into()
-        }
-        SExp::Pair(left, right) => {
-            let mut sha256 = Sha256::new();
-            sha256.update(&[2_u8]);
-            sha256.update(tree_hash(a, left));
-            sha256.update(tree_hash(a, right));
-            sha256.finalize().into()
+pub fn tree_hash(a: &Allocator, node: NodePtr) -> [u8; 32] {
+    let mut hashes: Vec<[u8; 32]> = vec![];
+    let mut ops = vec![TreeOp::SExp(node)];
+
+    while !ops.is_empty() {
+        match ops.pop().unwrap() {
+            TreeOp::SExp(node) => match a.sexp(node) {
+                SExp::Atom(atom) => {
+                    let mut sha256 = Sha256::new();
+                    sha256.update([1_u8]);
+                    sha256.update(a.buf(&atom));
+                    hashes.push(sha256.finalize().into())
+                }
+                SExp::Pair(left, right) => {
+                    ops.push(TreeOp::Cons);
+                    ops.push(TreeOp::SExp(left));
+                    ops.push(TreeOp::SExp(right));
+                }
+            },
+            TreeOp::Cons => {
+                let mut sha256 = Sha256::new();
+                sha256.update([2_u8]);
+                sha256.update(hashes.pop().unwrap());
```

Comment:
As a policy, better to use `.expect` rather than `.unwrap`. Shouldn't matter here though since unwrap can't really fail.

Reply:
good point